File: java/src/main/java/org/fluro/Checkout.java
package org.fluro;

import java.util.ArrayList;
import java.util.List;

class Checkout {
    private final List<Character> basket = new ArrayList<>();
    private final Stock stock;

    public Checkout(Stock stock) {
        this.stock = stock;
    }

    public int total() {
        return basket.stream().mapToInt((sku) -> stock.unitPrice(sku)).sum();
    }

    public Checkout scanItem(char sku) {
        basket.add(sku);
        return this;
    }
}

File: kotlin/src/main/kotlin/org/fluro/Checkout.kt
package org.fluro

data class Item(val sku: Char, val priceInPence: Int)

data class Stock(private val items: Set<Item>) {
    fun unitPrice(sku: Char) =
        items.firstOrNull { it.sku == sku }?.priceInPence
            ?: throw RuntimeException("sku $sku not found")
}

class Checkout(private val stock: Stock) {
    private val basket = mutableListOf<Char>()

    fun total(): Int = basket.sumOf { stock.unitPrice(it) }

    fun scanItem(sku: Char): Checkout {
        basket.add(sku)
        return this
    }
}
